File: python/leetcode/array/1169_invalid_transactions.py
"""
1169. Invalid Transactions (Easy)

A transaction is possibly invalid if:

the amount exceeds $1000, or;
if it occurs within (and including) 60 minutes of another transaction with the same name in a different city.
Each transaction string transactions[i] consists of comma separated values representing the name, time (in minutes), amount, and city of the transaction.

Given a list of transactions, return a list of transactions that are possibly invalid.  You may return the answer in any order.

Example 1:

Input: transactions = ["alice,20,800,mtv","alice,50,100,beijing"]
Output: ["alice,20,800,mtv","alice,50,100,beijing"]
Explanation: The first transaction is invalid because the second transaction occurs within a difference of 60 minutes, have the same name and is in a different city. Similarly the second one is invalid too.
Example 2:

Input: transactions = ["alice,20,800,mtv","alice,50,1200,mtv"]
Output: ["alice,50,1200,mtv"]
Example 3:

Input: transactions = ["alice,20,800,mtv","bob,50,1200,mtv"]
Output: ["bob,50,1200,mtv"]

Constraints:

transactions.length <= 1000
Each transactions[i] takes the form "{name},{time},{amount},{city}"
Each {name} and {city} consist of lowercase English letters, and have lengths between 1 and 10.
Each {time} consist of digits, and represent an integer between 0 and 1000.
Each {amount} consist of digits, and represent an integer between 0 and 2000.
"""

class Solution(object):
    def invalidTransactions(self, transactions):
        """
        :type transactions: List[str]
        :rtype: List[str]
        """
        memo = {} # name: (t, v, city)
        trans = []
        for tran in transactions:
            tlist = tran.split(',')
            key, t, v, city = tlist[0], int(tlist[1]), int(tlist[2]), tlist[3]
            trans.append((t, key, v, city))
        trans.sort()
        
        for tran in trans:
            t, k, v, city = tran
            if k not in memo:
                memo[k] = []
            memo[k].append((t, v, city))
        
        result = set()
        for k in memo.keys():
            last_id = 0
            for i, items in enumerate(memo[k]): # (t, v, city)
                t, v, city = items
                if v > 1000:
                    result.add(",".join([k,str(t),str(v),city]))
                flag = False
                for last_id in range(i):
                    t1, v1, city1 = memo[k][last_id]
                    if t - t1 <= 60 and city1 != city:
                        result.add(",".join([k,str(t1),str(v1),city1]))
                        flag = True

                if flag:
                    result.add(",".join([k,str(t),str(v),city]))
        result = list(result)
        result.sort()
        print(len(result))
        return result


if __name__ == "__main__":
    a = Solution()
    print(a.invalidTransactions(["alice,20,800,mtv","alice,50,100,beijing"]))
    print(a.invalidTransactions(["alice,20,800,mtv","alice,50,1200,mtv"]))
    print(a.invalidTransactions(["alice,20,800,mtv","bob,50,1200,mtv"]))

File: cc/leetcode/array/1169_invalid_transactions.cc
class Solution {
public:
    void parse(string& s1, vector<string>& v2) {
        int i=0, j=0, vid=0;
        while (i < s1.size()) {
            j = i;
            while (j < s1.size() && s1[j] != ',')
                j++;
            v2[vid++] = s1.substr(i, j-i);
            i = j+1;
        }
    }
    
    vector<string> invalidTransactions(vector<string>& transactions) {
        unordered_map<string, vector<tuple<string, string, string>>> memo;
        vector<vector<string>> trans;
        for (auto& tran : transactions) {
            // istringstream ss(tran);
            vector<string> s(4, "");
            // int i = 0;
            parse(tran, s);
            trans.push_back(s);
        }
        // sort
        sort(trans.begin(), trans.end(), [](const vector<string>& l, const vector<string>& r) {
            return stoi(l[1]) < stoi(r[1]);
        });
        // parse
        for (auto& tran : trans) {
            memo[tran[0]].push_back({tran[1], tran[2], tran[3]});
        }
        unordered_set<string> res;
        for (auto& kv : memo) {
            // kv: (k, [(t1,v1,city1), (t2,v2,city2), ...])
            int i, j;
            string k = kv.first;
            auto& itemlist = kv.second;
            for (i = 0; i < itemlist.size(); ++i)  {
                int t = stoi(get<0>(itemlist[i]));
                int v = stoi(get<1>(itemlist[i]));
                string& city = get<2>(itemlist[i]);
                
                bool flag = false;
                for (j = 0; j < i; ++j) {
                    int t1 = stoi(get<0>(itemlist[j]));
                    string& city1 = get<2>(itemlist[j]);
                    if (t-t1<=60 && city1 != city) {
                        string tmp = k + ',' + get<0>(itemlist[j]);
                        tmp += ',' + get<1>(itemlist[j]);
                        tmp += ',' + get<2>(itemlist[j]);
                        flag = true;
                        res.insert(tmp);
                    }
                }
                if (flag || v > 1000) {
                    string tmp = k + ',' + get<0>(itemlist[i]);
                    tmp += ',' + get<1>(itemlist[i]);
                    tmp += ',' + get<2>(itemlist[i]);
                    res.insert(tmp);
                }
            }
        }
        
        vector<string> result;
        for (string tmp : res)
            result.push_back(tmp);
        sort(result.begin(), result.end());
        return result;
    }
};
